felix/bpf/tc/attach.go

```diff
@@ -117,18 +166,56 @@ func (ap *AttachPoint) AttachProgram() error {
 		return err
 	}
 	logCxt.Info("Program attached to TC.")
+	// Remove any tcx program.
+	err = ap.detachTcxProgram()
+	if err != nil {
+		logCxt.Warnf("error removing tcx program from %s", err)
+	}
 	return nil
 }
 
-func (ap *AttachPoint) DetachProgram() error {
+func (ap *AttachPoint) progPinPath() string {
+	return path.Join(bpfdefs.TcxPinDir, fmt.Sprintf("%s_%s", strings.Replace(ap.Iface, ".", "", -1), ap.Hook))
+}
+
+func (ap *AttachPoint) detachTcxProgram() error {
+	progPinPath := ap.progPinPath()
+	link, err := libbpf.OpenLink(progPinPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return fmt.Errorf("error opening link %s:%w", progPinPath, err)
+	}
+	defer link.Close()
+	err = link.Detach()
+	if err != nil {
+		return fmt.Errorf("error detaching link %s:%w", progPinPath, err)
+	}
+	os.Remove(progPinPath)
+	return nil
+}
+
+func (ap *AttachPoint) detachTcProgram() error {
 	progsToClean, err := ListAttachedPrograms(ap.Iface, ap.Hook.String(), true)
 	if err != nil {
 		return err
 	}
-
 	return ap.detachPrograms(progsToClean)
 }
 
+func (ap *AttachPoint) DetachProgram() error {
+	err := ap.detachTcxProgram()
+	if err != nil {
```

Comment:
should we perhaps propagate the enoexists up to here and resolve it here so it is a bit more obvious that we are not just going to spam logs in tc mode?

Reply:
```detachTcxProgram``` is the main API to remove tcx pins and it is called when using the `tc` mode. I feel its better this way.

Reply:
libbpf throws the below warning ```failed to open link at``` when the link we are trying to open does not exist. Hence I changed to test if the pin path exists and return nil without opening the link.